gc/src/trace.rs

```diff
@@ -109,7 +113,71 @@ macro_rules! simple_empty_finalize_trace {
     }
 }
 
-simple_empty_finalize_trace![usize, bool, i8, u8, i16, u16, i32, u32, i64, u64, f32, f64, String];
+simple_empty_finalize_trace![(), isize, usize, bool, i8, u8, i16, u16, i32, u32, i64, u64, f32, f64, char, String, Path, PathBuf];
+
+macro_rules! fn_finalize_trace_one {
+    ($ty:ty $(,$args:ident)*) => {
+        impl<Ret $(,$args)*> Finalize for $ty {}
+        unsafe impl<Ret $(,$args)*> Trace for $ty { unsafe_empty_trace!(); }
+    }
+}
+macro_rules! fn_finalize_trace_group {
+    () => {
+        fn_finalize_trace_one!(extern "Rust" fn () -> Ret);
+        fn_finalize_trace_one!(extern "C" fn () -> Ret);
+        fn_finalize_trace_one!(unsafe extern "Rust" fn () -> Ret);
+        fn_finalize_trace_one!(unsafe extern "C" fn () -> Ret);
+    };
+    ($($args:ident),*) => {
+        fn_finalize_trace_one!(extern "Rust" fn ($($args),*) -> Ret, $($args),*);
+        fn_finalize_trace_one!(extern "C" fn ($($args),*) -> Ret, $($args),*);
+        fn_finalize_trace_one!(extern "C" fn ($($args),*, ...) -> Ret, $($args),*);
+        fn_finalize_trace_one!(unsafe extern "Rust" fn ($($args),*) -> Ret, $($args),*);
+        fn_finalize_trace_one!(unsafe extern "C" fn ($($args),*) -> Ret, $($args),*);
+        fn_finalize_trace_one!(unsafe extern "C" fn ($($args),*, ...) -> Ret, $($args),*);
+    }
+}
+
+macro_rules! tuple_finalize_trace {
+    () => {};
+    ($($args:ident),*) => {
+        impl<$($args),*> Finalize for ($($args,)*) {}
+        unsafe impl<$($args: $crate::Trace),*> Trace for ($($args,)*) {
+            custom_trace!(this, {
+                #[allow(non_snake_case, unused_unsafe)]
+                fn avoid_lints<$($args: $crate::Trace),*>(&($(ref $args,)*): &($($args,)*)) {
+                    unsafe { $(mark($args);)* }
+                }
+                avoid_lints(this)
+            });
+        }
+    }
+}
+
+macro_rules! table_based_finalize_trace {
```

Comment:
Should this be `tuple_based`?

Reply:
Changed to clarify; it's both tuples and functions, so I was like, "It's a table of types."

I'm renaming it to `type_arg_tuple_based_finalized_trace_impls`, unless you think that's too long.